### ISSUE TITLE
Crash when trying to log a maintenance log (beta version)

### ISSUE THREAD

**Issue by Lineflyer** [cgeo/cgeo#13343]
### Describe your problem!

User with beta version 2022.08.08 on support mail reports, that every time he tries to log a "need maintenance" for GC4JR1B the app crashes. Did not try to reproduce yet.
The log contains the crash pasted below.

### How to reproduce?

- Log GC4JR1B with "Needs maintenance"

### Actual result after these steps?

```
8-20 08:28:44.917 30646 30646 E AndroidRuntime: FATAL EXCEPTION: main
08-20 08:28:44.917 30646 30646 E AndroidRuntime: Process: cgeo.geocaching, PID: 30646
08-20 08:28:44.917 30646 30646 E AndroidRuntime: java.lang.IndexOutOfBoundsException: Index: 2, Size: 2
08-20 08:28:44.917 30646 30646 E AndroidRuntime: 	at java.util.ArrayList.get(ArrayList.java:437)
08-20 08:28:44.917 30646 30646 E AndroidRuntime: 	at cgeo.geocaching.ui.dialog.SimpleDialog.lambda$selectSingleGrouped$1(SimpleDialog.java:340)
08-20 08:28:44.917 30646 30646 E AndroidRuntime: 	at cgeo.geocaching.ui.dialog.SimpleDialog.lambda$selectSingleGrouped$1$SimpleDialog(Unknown Source:0)
08-20 08:28:44.917 30646 30646 E AndroidRuntime: 	at cgeo.geocaching.ui.dialog.-$$Lambda$SimpleDialog$_BEHauIquqxXzy2ZLeBXkgZWqss.onClick(Unknown Source:14)
08-20 08:28:44.917 30646 30646 E AndroidRuntime: 	at androidx.appcompat.app.AlertController$AlertParams$3.onItemClick(AlertController.java:1068)
08-20 08:28:44.917 30646 30646 E AndroidRuntime: 	at android.widget.AdapterView.performItemClick(AdapterView.java:330)
08-20 08:28:44.917 30646 30646 E AndroidRuntime: 	at android.widget.AbsListView.performItemClick(AbsListView.java:1197)
08-20 08:28:44.917 30646 30646 E AndroidRuntime: 	at android.widget.AbsListView$PerformClick.run(AbsListView.java:3216)
08-20 08:28:44.917 30646 30646 E AndroidRuntime: 	at android.widget.AbsListView$3.run(AbsListView.java:4192)
08-20 08:28:44.917 30646 30646 E AndroidRuntime: 	at android.os.Handler.handleCallback(Handler.java:938)
08-20 08:28:44.917 30646 30646 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:99)
08-20 08:28:44.917 30646 30646 E AndroidRuntime: 	at android.os.Looper.loopOnce(Looper.java:226)
08-20 08:28:44.917 30646 30646 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:329)
08-20 08:28:44.917 30646 30646 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:8058)
08-20 08:28:44.917 30646 30646 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
08-20 08:28:44.917 30646 30646 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
08-20 08:28:44.917 30646 30646 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1026)
08-20 08:29:00.108  8704  8704 E AndroidRuntime: FATAL EXCEPTION: main
08-20 08:29:00.108  8704  8704 E AndroidRuntime: Process: cgeo.geocaching, PID: 8704
08-20 08:29:00.108  8704  8704 E AndroidRuntime: java.lang.IndexOutOfBoundsException: Index: 2, Size: 2
08-20 08:29:00.108  8704  8704 E AndroidRuntime: 	at java.util.ArrayList.get(ArrayList.java:437)
08-20 08:29:00.108  8704  8704 E AndroidRuntime: 	at cgeo.geocaching.ui.dialog.SimpleDialog.lambda$selectSingleGrouped$1(SimpleDialog.java:340)
08-20 08:29:00.108  8704  8704 E AndroidRuntime: 	at cgeo.geocaching.ui.dialog.SimpleDialog.lambda$selectSingleGrouped$1$SimpleDialog(Unknown Source:0)
08-20 08:29:00.108  8704  8704 E AndroidRuntime: 	at cgeo.geocaching.ui.dialog.-$$Lambda$SimpleDialog$_BEHauIquqxXzy2ZLeBXkgZWqss.onClick(Unknown Source:14)
08-20 08:29:00.108  8704  8704 E AndroidRuntime: 	at androidx.appcompat.app.AlertController$AlertParams$3.onItemClick(AlertController.java:1068)
08-20 08:29:00.108  8704  8704 E AndroidRuntime: 	at android.widget.AdapterView.performItemClick(AdapterView.java:330)
08-20 08:29:00.108  8704  8704 E AndroidRuntime: 	at android.widget.AbsListView.performItemClick(AbsListView.java:1197)
08-20 08:29:00.108  8704  8704 E AndroidRuntime: 	at android.widget.AbsListView$PerformClick.run(AbsListView.java:3216)
08-20 08:29:00.108  8704  8704 E AndroidRuntime: 	at android.widget.AbsListView$3.run(AbsListView.java:4192)
08-20 08:29:00.108  8704  8704 E AndroidRuntime: 	at android.os.Handler.handleCallback(Handler.java:938)
08-20 08:29:00.108  8704  8704 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:99)
08-20 08:29:00.108  8704  8704 E AndroidRuntime: 	at android.os.Looper.loopOnce(Looper.java:226)
08-20 08:29:00.108  8704  8704 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:329)
08-20 08:29:00.108  8704  8704 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:8058)
08-20 08:29:00.108  8704  8704 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
08-20 08:29:00.108  8704  8704 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
08-20 08:29:00.108  8704  8704 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1026)
08-20 08:30:23.306  8905  8905 E AndroidRuntime: FATAL EXCEPTION: main
08-20 08:30:23.306  8905  8905 E AndroidRuntime: Process: cgeo.geocaching, PID: 8905
08-20 08:30:23.306  8905  8905 E AndroidRuntime: java.lang.IndexOutOfBoundsException: Index: 2, Size: 2
08-20 08:30:23.306  8905  8905 E AndroidRuntime: 	at java.util.ArrayList.get(ArrayList.java:437)
08-20 08:30:23.306  8905  8905 E AndroidRuntime: 	at cgeo.geocaching.ui.dialog.SimpleDialog.lambda$selectSingleGrouped$1(SimpleDialog.java:340)
08-20 08:30:23.306  8905  8905 E AndroidRuntime: 	at cgeo.geocaching.ui.dialog.SimpleDialog.lambda$selectSingleGrouped$1$SimpleDialog(Unknown Source:0)
08-20 08:30:23.306  8905  8905 E AndroidRuntime: 	at cgeo.geocaching.ui.dialog.-$$Lambda$SimpleDialog$_BEHauIquqxXzy2ZLeBXkgZWqss.onClick(Unknown Source:14)
08-20 08:30:23.306  8905  8905 E AndroidRuntime: 	at androidx.appcompat.app.AlertController$AlertParams$3.onItemClick(AlertController.java:1068)
08-20 08:30:23.306  8905  8905 E AndroidRuntime: 	at android.widget.AdapterView.performItemClick(AdapterView.java:330)
08-20 08:30:23.306  8905  8905 E AndroidRuntime: 	at android.widget.AbsListView.performItemClick(AbsListView.java:1197)
08-20 08:30:23.306  8905  8905 E AndroidRuntime: 	at android.widget.AbsListView$PerformClick.run(AbsListView.java:3216)
08-20 08:30:23.306  8905  8905 E AndroidRuntime: 	at android.widget.AbsListView$3.run(AbsListView.java:4192)
08-20 08:30:23.306  8905  8905 E AndroidRuntime: 	at android.os.Handler.handleCallback(Handler.java:938)
08-20 08:30:23.306  8905  8905 E AndroidRuntime: 	at android.os.Handler.dispatchMessage(Handler.java:99)
08-20 08:30:23.306  8905  8905 E AndroidRuntime: 	at android.os.Looper.loopOnce(Looper.java:226)
08-20 08:30:23.306  8905  8905 E AndroidRuntime: 	at android.os.Looper.loop(Looper.java:329)
08-20 08:30:23.306  8905  8905 E AndroidRuntime: 	at android.app.ActivityThread.main(ActivityThread.java:8058)
08-20 08:30:23.306  8905  8905 E AndroidRuntime: 	at java.lang.reflect.Method.invoke(Native Method)
08-20 08:30:23.306  8905  8905 E AndroidRuntime: 	at com.android.internal.os.RuntimeInit$MethodAndArgsCaller.run(RuntimeInit.java:548)
08-20 08:30:23.306  8905  8905 E AndroidRuntime: 	at com.android.internal.os.ZygoteInit.main(ZygoteInit.java:1026)
```

### Expected result after these steps?

Normal operation

### Reproducible

Unclear

### c:geo Version

2022.08.08

### System information

```text
## System information

c:geo version: 2022.08.08

Device:
-------
- Device type: moto g(30) (caprip_retailen, motorola)
- Available processors: 8
- Android version: 12
- Android build: S0RCS32.41-10-9-2-4
- Screen resolution: 720x1473px (376x764dp)
- Pixel density: 1.9125
- System font scale: 1.0 / used scale: 1.0
- Sailfish OS detected: false
- Google Play services: enabled - 22.26.15 (190400-461192076)

Sensor and location:
-------
- Low power mode: inactive
- Compass capabilities: no
- Rotation vector sensor: absent
- Orientation sensor: absent
- Magnetometer & Accelerometer sensor: absent
- Direction sensor used: magnetometer & accelerometer

Program settings:
-------
- Settings: v9, Count:178
- Set language: fr_FR (system default)
- System date format: dd/MM/y
- Time zone: GMT+02:00
- Debug mode active: no
- Last backup: 14 août, 01:23
- Routing mode: Off
- Live map mode: true
- OSM multi-threading: true / threads: 4
- Map: OpenStreetMap.org
  - Id: cgeo.geocaching.maps.mapsforge.MapsforgeMapProvider$OsmMapSource
  - Atts: OpenStreetMap DE, map data OpenStreetMap contributors
  - Theme: none

Filters:
-------
- Hide waypoints: visited
- LIVE: Statut:Exclure les caches désactivées,Exclure les caches archivées ([:inconclusive=false:advanced=true]AND(status:exclude_disabled:exclude_archived))
- OFFLINE: Aucun ([:inconclusive=false:advanced=true])

Stored Filters (#0):

Services:
-------
- Geocaching sites enabled:
   geocaching.com: Logged in (Connecté) / PREMIUM
   Geocaching.com Adventure Lab
- Geocaching.com date format: dd/MM/yyyy
- Last login error on geocaching.com: communication error (2022-08-17 17-35)
- Last successful login on geocaching.com: 2022-08-20 09-27
- Routing: external / BRouter installed: false
- Installed c:geo plugins:  none

Permissions & paths:
-------
- Fine location permission: granted
- Write external storage permission: granted
- System internal c:geo dir: /data/user/0/cgeo.geocaching (37,7 GB free) v3 internal isDir(31 entries)
- Legacy User storage c:geo dir: /data/user/0/cgeo.geocaching (37,7 GB free) v3 internal isDir(31 entries)
- Geocache data: /storage/emulated/0/Android/data/cgeo.geocaching/files/GeocacheData (37,7 GB free) v3 external non-removable isDir(2152 entries)
- Internal theme sync (is turned off): /data/user/0/cgeo.geocaching/MapThemeData (37,7 GB free) v3 internal isDir(0 entries)
- Public Folders: #11
  - BASE: /cgeo (User-Defined)[/cgeo[DOCUMENT#0:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo, Av:true, files:>=15, dirs:>=14, totalFileSize:>=3,1 GB, free space: 37,7 GB, files on device: 13511677)
  - OFFLINE_MAPS: /cgeo/maps (Default)[/cgeo/maps[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/maps]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Fmaps, Av:true, files:11, dirs:1, totalFileSize:3,1 GB, free space: 37,7 GB, files on device: 13511677)
  - OFFLINE_MAP_THEMES: /cgeo/maps/_themes (Default)[/cgeo/maps/_themes[PERSISTABLE_FOLDER(OFFLINE_MAPS)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/maps/_themes]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Fmaps%2F_themes, Av:true, files:2, dirs:0, totalFileSize:504,6 KB, free space: 37,7 GB, files on device: 13511677)
  - LOGFILES: /cgeo/logfiles (Default)[/cgeo/logfiles[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/logfiles]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Flogfiles, Av:true, files:3, dirs:0, totalFileSize:34,3 KB, free space: 37,7 GB, files on device: 13511677)
  - GPX: /cgeo/gpx (Default)[/cgeo/gpx[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/gpx]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Fgpx, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 37,7 GB, files on device: 13511677)
  - BACKUP: /cgeo/backup (Default)[/cgeo/backup[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/backup]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Fbackup, Av:true, files:>=10, dirs:>=6, totalFileSize:>=552,7 MB, free space: 37,7 GB, files on device: 13511677)
  - FIELD_NOTES: /cgeo/field-notes (Default)[/cgeo/field-notes[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/field-notes]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Ffield-notes, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 37,7 GB, files on device: 13511677)
  - SPOILER_IMAGES: /cgeo/GeocachePhotos (Default)[/cgeo/GeocachePhotos[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/GeocachePhotos]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2FGeocachePhotos, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 37,7 GB, files on device: 13511677)
  - ROUTING_BASE: /cgeo/routing (Default)[/cgeo/routing[PERSISTABLE_FOLDER(BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/routing]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Frouting, Av:true, files:7, dirs:1, totalFileSize:81,2 KB, free space: 37,7 GB, files on device: 13511677)
  - ROUTING_TILES: /cgeo/routing/segments4 (Default)[/cgeo/routing/segments4[PERSISTABLE_FOLDER(ROUTING_BASE)#1:p-content://com.android.externalstorage.documents/tree/primary%3Acgeo::/routing/segments4]] (Uri: content://com.android.externalstorage.documents/tree/primary%3Acgeo/document/primary%3Acgeo%2Frouting%2Fsegments4, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 37,7 GB, files on device: 13511677)
  - TEST_FOLDER: [Legacy]/data/user/0/cgeo.geocaching/files/unittest (Default)[/data/user/0/cgeo.geocaching/files/unittest[FILE#1:p-file:///data/user/0/cgeo.geocaching/files::/unittest]] (Uri: file:///data/user/0/cgeo.geocaching/files/unittest, Av:true, files:0, dirs:0, totalFileSize:0 B, free space: 37,7 GB, files on device: -1)
- Map render theme path:
- PersistedDocumentUris: #1
- UNUSED_TRACK: null
- Persisted Uri Permissions: #1
  - content://com.android.externalstorage.documents/tree/primary%3Acgeo (17 juil. 2021, 21:44):RW
- Database: /data/user/0/cgeo.geocaching/databases/data (v99, Size:121,8 MB) on system internal storage

--- End of system information ---
```

### Additional Information

Support ticket 253213

**Comment by moving-bits** [comment]
May be related to a ticket on support stating that no "needs maintenance" option is available for the user any more. I've debugged a bit into it. c:geo parses available log types by looking for strings like:
```
 window.logSettings.logTypes.push({"Value":2,"Description":"Found It","IsRealtimeOnly":false}); 
 window.logSettings.logTypes.push({"Value":3,"Description":"Didn\u0027t Find It","IsRealtimeOnly":false}); 
 window.logSettings.logTypes.push({"Value":4,"Description":"Write note","IsRealtimeOnly":false}); 
```
For some reason yet unknown to me there's no entry for "needs maintenance" any more for some caches.

**Comment by moving-bits** [comment]
Setting priority to "High" as we are receiving further tickets on support that they cannot log "needs maintenance" any more.

**Comment by Lineflyer** [comment]
It seems Groundspeak removed the possibility to explicitly log a "Needs maintenance". You can also see that on the website, where for logging only Found/Not found/Note are available.

If you need to log a needs maintenance you have to use the "problem report" section below the log text (in c:geo same as on the website). 

Not sure how the crash above might happen. Maybe if users offline log "Needs maintenance" and later try to convert it to an online log which causes a mismatch in available log types?

**Comment by fm-sys** [comment]
Maybe we currently submit the NM log separately?

**Comment by ztNFny** [comment]
> For some reason yet unknown to me there's no entry for "needs maintenance" any more for some caches.

Do you have a GC code where NM is not possible or is it happening randomly? I couldn't find one without.

> It seems Groundspeak removed the possibility to explicitly log a "Needs maintenance". You can also see that on the website, where for logging only Found/Not found/Note are available.

We've been using the new log forms "Report problem" feature for a long time. NM logs have only been possible via API for quite a while, so that shouldn't be it...

**Comment by moving-bits** [comment]
> Do you have a GC code where NM is not possible or is it happening randomly? I couldn't find one without.

For me it was/is happening consistently, using a build from `master` from a couple of days ago. The result when you tap "log visit" is like this:

![image](https://user-images.githubusercontent.com/3754370/188226469-7749f6f1-a0b9-4ca5-ac24-fdbd1514c365.png)

> > It seems Groundspeak removed the possibility to explicitly log a "Needs maintenance". You can also see that on the website, where for logging only Found/Not found/Note are available.
> 
> We've been using the new log forms "Report problem" feature for a long time. NM logs have only been possible via API for quite a while, so that shouldn't be it...

My observation during a short debugging session was, that when opening the log visit activity, available log types are determined by the connector delivering a hard-coded list - which contains "log maintenance" for gc connector. This list is then updated by parsing the cache's source code which contains available log types as mentioned in https://github.com/cgeo/cgeo/issues/13343#issuecomment-1229026124.

I guess the crash happens like this:
- open a cache, select "log visit", and directly select a "needs maintenance" (while the cache's source is still being analyzed or when opening an offline log)
- when the list of available types got updated meanwhile, the index for "needs maintenance" is no longer available => crash

**Comment by ztNFny** [comment]
> > Do you have a GC code where NM is not possible or is it happening randomly? I couldn't find one without.
> 
> For me it was/is happening consistently, using a build from `master` from a couple of days ago. The result when you tap "log visit" is like this:
> 
> ![image](https://user-images.githubusercontent.com/3754370/188226469-7749f6f1-a0b9-4ca5-ac24-fdbd1514c365.png)

That looks like expected. NM/NA is availble in the "report problem" dropdown below the log text field.

> > > It seems Groundspeak removed the possibility to explicitly log a "Needs maintenance". You can also see that on the website, where for logging only Found/Not found/Note are available.
> > 
> > 
> > We've been using the new log forms "Report problem" feature for a long time. NM logs have only been possible via API for quite a while, so that shouldn't be it...
> 
> My observation during a short debugging session was, that when opening the log visit activity, available log types are determined by the connector delivering a hard-coded list - which contains "log maintenance" for gc connector. 

If the list still contains NM/NA they should be removed as they haven't been available in a long time (since the new log form - 2 years or so?)

**Comment by moving-bits** [comment]
For current release 2022.08.21, a "needs maintenance" is actually being offered as an option. The menu looks like this for the same cache:

![image](https://user-images.githubusercontent.com/3754370/188230969-9d519c92-950b-46cc-907c-2a73c2988194.png)

**Comment by ztNFny** [comment]
> For current release 2022.08.21, a "needs maintenance" is actually being offered as an option. The menu looks like this for the same cache:
> 
> ![image](https://user-images.githubusercontent.com/3754370/188230969-9d519c92-950b-46cc-907c-2a73c2988194.png)

I haven't had those 2 options for ages. You sure the log page has loaded for that screenshot?

**Comment by ztNFny** [comment]
Release 2022.08.21

Before loading log page: 5 options
After loading (visible by TBs being loaded): 3 options

It's been like this since the new log page

**Comment by moving-bits** [comment]
> I haven't had those 2 options for ages. You sure the log page has loaded for that screenshot?

probably not. I was testing in the emulator, and retrieving the log data may take enough time for me being able to hit the "log type" selection meanwhile.

If that's actually the cause, we should suppress the default "needs maintenance" option at least, IMHO.

**Comment by ztNFny** [comment]
And the needs archive as well

**Comment by moving-bits** [comment]
Only thing I wonder is why that has come up with our current release - what has changed there in terms of logging?
(Just have seen another ticket on support stating c:geo crashes on logging "needs maintenance".)

**Comment by moving-bits** [comment]
See #13384 for removal of nm/na from default log types for gc connector

**Comment by Lineflyer** [comment]
You are perfectly right regarding the automatic adaption of available log types before and after the log page has been loaded.
So those NA/NM should probably be removed from the connectors default list.

Please be aware to correctly handle the "quick offlline log" feature in combination with that change:
Up to now you can produce a NM or NA "quick offline log". Those should probably be auto converted to a "Write note" with the corresponding "problem report".

**Comment by ztNFny** [comment]
> Those should probably be auto converted to a "Write note" with the corresponding "problem report".

I'm 90% sure that's already the case.

**Comment by moving-bits** [comment]
> > Those should probably be auto converted to a "Write note" with the corresponding "problem report".
> 
> I'm 90% sure that's already the case.

That's indeed in the code (`LoggingUI.java`), but since my removal of NM/NA from default options for GC connector, those options are no longer shown in the quick logging menu. I'll look into it.

**Comment by moving-bits** [comment]
Should be fixed with 2022.09.04. Have not seen any new issues on support, so let's close here.
